src/main/scala/scorex/core/network/peer/SessionIdPeerFeature.scala

```diff
@@ -0,0 +1,39 @@
+package scorex.core.network.peer
+
+import java.net.{InetAddress, InetSocketAddress}
+
+import scorex.core.network.PeerFeature
+import scorex.core.network.PeerFeature.Id
+import scorex.util.serialization._
+import scorex.core.serialization.ScorexSerializer
+import scorex.util.Extensions._
+
+/**
+  * This peer feature allows to more reliably detect connections to self node and connections from other network
+  * @param networkMagic network magic bytes (taken from settings)
+  * @param sessionId randomly generated 64-bit session identifier
+  */
+case class SessionIdPeerFeature(networkMagic:Array[Byte], sessionId:Long = scala.util.Random.nextLong()) extends PeerFeature {
+  override type M = SessionIdPeerFeature
+  override val featureId: Id = SessionIdPeerFeature.featureId
+
+  override def serializer: SessionIdPeerFeatureSerializer.type = SessionIdPeerFeatureSerializer
+}
+
+object SessionIdPeerFeature {
+  val featureId: Id = 3: Byte
+}
+
+object SessionIdPeerFeatureSerializer extends ScorexSerializer[SessionIdPeerFeature] {
+  override def serialize(obj: SessionIdPeerFeature, w: Writer): Unit = {
+    w.putInt(obj.networkMagic.size)
+    w.putBytes(obj.networkMagic)
+    w.putLong(obj.sessionId)
+  }
+
+  override def parse(r: Reader): SessionIdPeerFeature = {
+    val networkMagic = r.getBytes(r.getInt())
```

Comment:
better to use putUInt / getUInt

Reply:
getUInt() returns Long and can not be passed to getBytes

Reply:
Actually, network magic is always about 4 bytes, so no need to pass this value